src/Database.h

```diff
@@ -0,0 +1,50 @@
+#ifndef _DATABASE_H_
+#define _DATABASE_H_
+
+#include "SURLCreator.h"
+#include <unordered_map>
+#include <mutex>
+
+// Database save the look-up table to link each SURL to there original URL
+// public implement: create, query, insert
+
+class Database
+{
+public:
+	Database();
+
+	// Input a rawURL, return SURL if create is successful. Return "" if create fail.
+	// Encode the rawURL to SURL and check whether it exists.
+	// If it exists update expire time, else create the URL.
+	std::string CreateSURL(std::string rawURL);
+
+	// Query whether URL exists. Return SURL if SURL exists. Return "" if SURL doesn't exist.
+	std::string QuerySURL(std::string surl);
+
+	// lock the mutex when it is inserting.
+	// There is only one inserting in the same time.
+	bool InsertSURL(SURL surl);
+private:
+
+	// If there is a rawURL which encoding to a SURL that exists in LUT,
+	// Update its expire time and don't need to create new one in LUT.
+	bool ModifySURL(SURL surl, std::string expire);
+
+	// It is for daily work. Check if the expire time of the SURL is over.
+	void CheckSURLExpiration();
+
+	// If the expire time of the SURL is over, erase if in lut.
+	bool ReleaseSURL();
+
+	// Encode the rawURL to the SURL and set the expire time to now.
+	SURL EncodeSURL(std::string rawURL);
+
+	// The look up table of the SURL. Use SURL as key to find the original url.
+	std::unordered_map <SURL, std::string> lut;
+
+	// Lock to prevent multiple user from modifying the SURL.
+	// There is only one insertion in the same time.
+	std::mutex writeLock;
```

Comment:
I'm not sure it's necessary of this part, I guess DB service usually provides the lock mechanism when doing update or similar works

Reply:
We use the stl unordered_map as look-up table. I checked for the cpp reference and it said that unordered_map is not thread safe. Therefore, I add the mutex to prevent from conflict.